src/utils/schedule.rs

```diff
@@ -0,0 +1,277 @@
+            let rrules = &rrules_with_header[1..];
+            let header_parts: Vec<&str> = header
+                .split(";DURATION:")
+                .filter(|s| !s.is_empty())
+                .collect();
+            if header_parts.len() != 2 {
+                return Err(());
+            }
+            let dtstart = header_parts[0];
+            let duration = header_parts[1];
+            let str = "DTSTART:".to_owned() + dtstart + "\n" + rrules.join("\n").as_str();
+            let rrule_set = RRuleSet::from_str(&str).unwrap();
+            recurrent_events.push(RecurrentEvent {
+                rrule_set,
+                duration: duration.to_string(),
+            });
+        }
+        Ok(Calendar {
+            events: recurrent_events,
+        })
+    }
+}
+
+impl Display for Calendar {
+    /// Formats `Calendar` into multiline string which can be stored in the database
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let err_msg = String::from("Error writing to string");
+        for event in &self.events {
+            writeln!(
+                f,
+                "DTSTART:{};DURATION:{}",
+                datetime_to_ical_format(event.rrule_set.get_dt_start()),
+                &event.duration
+            )
+            .expect(&err_msg);
+            for rrule in event.rrule_set.get_rrule() {
+                writeln!(f, "RRULE:{}", rrule).expect(&err_msg);
+            }
+            for rdate in event.rrule_set.get_rdate() {
+                writeln!(f, "RDATE:{}", datetime_to_ical_format(rdate)).expect(&err_msg);
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Calendar {
+    /// Wrapper implementation of rrule library's `all` method which also considers duration of the event
+    pub fn is_available_between(&self, start_time: DateTime<Tz>, end_time: DateTime<Tz>) -> bool {
+        // adjust start and end time by one second to make search inclusive of boundary values
+        let start_time = start_time + Duration::seconds(1);
+        let end_time = end_time - Duration::seconds(1);
+        for event in &self.events {
+            let duration = &event.duration;
+            // check standard rrule time - if event(block) start time is between two dates,
+            // then it will be found and time slot will be marked as not available
+            let (events, _) = &event
+                .rrule_set
+                .clone()
+                .after(start_time)
+                .before(end_time)
+                .all(1);
+            if !events.is_empty() {
+                return false;
+            }
+            let d = DurationParser::parse(duration).expect("Failed to parse duration");
+            let adjusted_start_time = start_time
+                - Duration::days(d.day as i64)
+                - Duration::hours(d.hour as i64)
+                - Duration::minutes(d.minute as i64)
+                - Duration::seconds(d.second as i64);
```

Comment:
why are we subtracting durations from start_time? I assume duration means how long it takes for a flight? So if start time is 8am, and duration is 30 minutes, then, this function here wants to check if the destination port is available at 8:30am, shouldn't it be start_time + duration?

However, if we are checking for the window, say 8am to 10am, we are only applying duration to the 8am. The port could be unavailable at 8:30. But if we depart at 9am, the port could be available at 9:30. Are we addressing this scenario in this function?

Reply:
In short: yes, the scenario is addressed.

Duration here is a duration of a "blocking event" in calendar. Not duration of the flight.

This function checks if the time window is fully available in the schedule of the vertiport or aircraft. It's very generic so basically it checks for availability time in any "calendar". (start_time and end time is a time window we check for availability - e.g. for the departure vertiport we check a period of loading and takeoff time, for aircraft we check the full flight time etc.)

Standard rrule library can only check if specific timestamp (without duration) is between 2 timestamps (start_time / end_time)
This is checked after this comment `// check standard rrule time - if event(block) start time is between two dates`

My extra implementation checks if (timestamp + duration) is fully between start_time and end_time. They way to do it is actually move start_time to the past by duration and see if original timestamp is within adjusted_start_time and end_time. 

There are tests verifying this works. `test_calendar_with_day_break` checks if the schedule between 11:30 and 12:30 is available. But there is a blocking 1hr event from 12:00 to 13:00 so the function returns false (not available)

Reply:
Ah, I see, that's very smart. Probably warrant writing an engineering blog about it haha.